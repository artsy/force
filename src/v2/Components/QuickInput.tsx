import { space } from "@artsy/palette"
import { fadeIn, fadeOut } from "v2/Assets/Animations"
import Colors from "v2/Assets/Colors"
import { garamond, unica } from "v2/Assets/Fonts"
import { Component } from "react";
import styled from "styled-components"
import { borderedInputMixin } from "./Mixins"

import { ExtractProps } from "v2/Utils/ExtractProps"
import { InputError } from "./Input"

export interface QuickInputProps extends ExtractProps<typeof InputComponent> {
  block?: boolean
  error?: string
  label?: string
  note?: string
  rightAddOn?: JSX.Element
  setTouched?: (fields: { [field: string]: boolean }) => void
  touchedOnChange?: boolean
}

export interface QuickInputState {
  focused: boolean
  value: string
}

/**
 * Quick input. Renders the label inside of the textbox.
 * @deprecated
 */
export class QuickInput extends Component<
  QuickInputProps,
  QuickInputState
> {
  state = {
    focused: false,
    value: (this.props.value as string) || "",
    touchedOnChange: true,
  }

  UNSAFE_componentWillReceiveProps(newProps) {
    if (this.props.name !== newProps.name) {
      this.setState({
        value: "",
      })
    }
  }

  onFocus = e => {
    this.setState({
      focused: true,
    })

    if (this.props.onFocus) {
      this.props.onFocus(e)
    }
  }

  onBlur = e => {
    if (this.props.setTouched) {
      // @ts-expect-error STRICT_NULL_CHECK
      this.props.setTouched({ [this.props.name]: true })
    }
    this.setState({
      focused: false,
    })

    if (this.props.onBlur) {
      this.props.onBlur(e)
    }
  }

  onChange = e => {
    if (this.props.touchedOnChange && this.props.setTouched) {
      // @ts-expect-error STRICT_NULL_CHECK
      this.props.setTouched({ [this.props.name]: true })
    }
    this.setState({
      value: e.currentTarget.value,
    })

    if (this.props.onChange) {
      this.props.onChange(e)
    }
  }

  render() {
    const {
      error,
      className,
      label,
      ref: _ref,
      onChange,
      setTouched,
      rightAddOn,
      note,
      ...newProps
    } = this.props
    const showLabel = (!!this.state.focused || !!this.state.value) && !!label

    return (
      <Container>
        <InputContainer
          hasLabel={!!label}
          hasError={!!error}
          className={this.state.focused ? "focused" : ""}
        >
          <Label out={!showLabel}>{label}</Label>
          <InputComponent
            {...newProps}
            onFocus={this.onFocus}
            onBlur={this.onBlur}
            onChange={this.onChange}
            value={this.state.value}
            showLabel={showLabel}
          />
          {rightAddOn}
        </InputContainer>
        {note && <InputNote>{note}</InputNote>}
        {error && <InputError>{error}</InputError>}
      </Container>
    )
  }
}

const Container = styled.div`
  padding-bottom: ${space(0.5)}px;
`

const InputComponent = styled.input`
  ${garamond("s17")};
  border: 0;
  outline: none;
  flex: 1;
  transition: all 0.25s;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 0 ${space(1)}px;
  line-height: initial;
  ${(props: { showLabel?: boolean }) =>
    props.showLabel && `padding: ${space(1)}px ${space(1)}px 0 ${space(1)}px`};
`

const InputContainer = styled.div<{
  hasLabel?: boolean
  hasError: boolean
}>`
  ${borderedInputMixin}
  margin-right: 0;
  margin-top: ${space(0.5)}px;
  margin-bottom: ${space(1)}px;
  display: flex;
  position: relative;
  height: ${p => (p.hasLabel ? `${space(4)}px` : `${space(2)}px`)};
  flex-direction: row;
  align-items: center;
  box-sizing: content-box;
`

const Label = styled.label<{ out: boolean }>`
  ${unica("s12", "medium")};
  position: absolute;
  left: ${space(1)}px;
  top: ${space(1)}px;
  visibility: ${p => (p.out ? "hidden" : "visible")};
  animation: ${p => (p.out ? fadeOut : fadeIn)} 0.2s linear;
  transition: visibility 0.2s linear;
  z-index: 1;
`

const InputNote = styled.div`
  ${unica("s12")};
  margin-top: ${space(1)}px;
  color: ${Colors.graySemibold};
  height: 16px;
`

export default QuickInput
