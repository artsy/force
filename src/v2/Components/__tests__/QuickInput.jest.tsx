import { mount } from "enzyme"
import "jest-styled-components"
import renderer from "react-test-renderer"

import QuickInput from "../QuickInput"

describe("QuickInput", () => {
  it("renders a QuickInput", () => {
    const wrapper = mount(<QuickInput />)

    expect(wrapper.find("input").length).toEqual(1)
  })

  it("renders a QuickInput with metadata", () => {
    const rightAddOn = <div>right side</div>

    const wrapper = mount(
      <QuickInput
        error="an error"
        label="some label"
        note="a note"
        rightAddOn={rightAddOn}
      />
    )

    const actual = wrapper.text()
    expect(actual).toContain("a note")
    expect(actual).toContain("an error")
    expect(actual).toContain("some label")
    expect(actual).toContain("right side")
  })

  it("renders an unfocused QuickInput as a snapshot", () => {
    const rightAddOn = <div>right side</div>
    const component = renderer
      .create(
        <QuickInput
          error="an error"
          label="some label"
          note="a note"
          rightAddOn={rightAddOn}
        />
      )
      .toJSON()
    expect(component).toMatchSnapshot()
  })

  it("renders a focused QuickInput as a snapshot", () => {
    const rightAddOn = <div>right side</div>
    const component = renderer.create(
      <QuickInput
        error="an error"
        label="some label"
        note="a note"
        rightAddOn={rightAddOn}
      />
    )

    const root = component.root
    const input = root.find(element => element.type === "input")
    input.props.onFocus()

    expect(component.toJSON()).toMatchSnapshot()
  })
})
