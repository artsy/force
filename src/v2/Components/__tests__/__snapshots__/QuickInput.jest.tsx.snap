// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QuickInput renders a focused QuickInput as a snapshot 1`] = `
.c5 {
  font-family: Unica77LLWebRegular,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  margin-top: 10px;
  color: #E82F1D;
  -webkit-transition: visibility 0.2s linear;
  transition: visibility 0.2s linear;
  -webkit-animation: hqnQVQ 0.25s linear;
  animation: hqnQVQ 0.25s linear;
  height: 16px;
}

.c0 {
  padding-bottom: 5px;
}

.c3 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 0;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 0 10px;
  line-height: initial;
  padding: 10px 10px 0 10px;
}

.c1 {
  padding: 10px;
  box-shadow: none;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 10px;
  resize: none;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 1px solid #f7625a;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 0;
  margin-top: 5px;
  margin-bottom: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 40px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: content-box;
}

.c1::-webkit-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::-moz-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:-ms-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:hover,
.c1:focus,
.c1.focused {
  border-color: #f7625a;
  outline: 0;
}

.c1:disabled {
  border: 2px dotted #e5e5e5;
}

.c2 {
  font-family: Unica77LLWebMedium,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  position: absolute;
  left: 10px;
  top: 10px;
  visibility: visible;
  -webkit-animation: jBcSpD 0.2s linear;
  animation: jBcSpD 0.2s linear;
  -webkit-transition: visibility 0.2s linear;
  transition: visibility 0.2s linear;
  z-index: 1;
}

.c4 {
  font-family: Unica77LLWebRegular,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  margin-top: 10px;
  color: #666666;
  height: 16px;
}

<div
  className="c0"
>
  <div
    className="c1 focused"
  >
    <label
      className="c2"
    >
      some label
    </label>
    <input
      className="c3"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value=""
    />
    <div>
      right side
    </div>
  </div>
  <div
    className="c4"
  >
    a note
  </div>
  <div
    className="c5"
  >
    an error
  </div>
</div>
`;

exports[`QuickInput renders an unfocused QuickInput as a snapshot 1`] = `
.c5 {
  font-family: Unica77LLWebRegular,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  margin-top: 10px;
  color: #E82F1D;
  -webkit-transition: visibility 0.2s linear;
  transition: visibility 0.2s linear;
  -webkit-animation: hqnQVQ 0.25s linear;
  animation: hqnQVQ 0.25s linear;
  height: 16px;
}

.c0 {
  padding-bottom: 5px;
}

.c3 {
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 0;
  outline: none;
  -webkit-flex: 1;
  -ms-flex: 1;
  flex: 1;
  -webkit-transition: all 0.25s;
  transition: all 0.25s;
  position: absolute;
  top: 0;
  left: 0;
  width: 100%;
  height: 100%;
  padding: 0 10px;
  line-height: initial;
}

.c1 {
  padding: 10px;
  box-shadow: none;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 10px;
  resize: none;
  display: block;
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
  font-family: 'Adobe Garamond W08','adobe-garamond-pro','AGaramondPro-Regular','Times New Roman',Times,serif;
  font-size: 17px;
  line-height: 1.1em;
  -webkit-font-smoothing: antialiased;
  border: 1px solid #f7625a;
  -webkit-transition: border-color 0.25s;
  transition: border-color 0.25s;
  margin-right: 0;
  margin-top: 5px;
  margin-bottom: 10px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  position: relative;
  height: 40px;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  box-sizing: content-box;
}

.c1::-webkit-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::-moz-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:-ms-input-placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1::placeholder {
  color: #cccccc;
  text-overflow: ellipsis;
  line-height: normal;
}

.c1:hover,
.c1:focus,
.c1.focused {
  border-color: #f7625a;
  outline: 0;
}

.c1:disabled {
  border: 2px dotted #e5e5e5;
}

.c2 {
  font-family: Unica77LLWebMedium,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  position: absolute;
  left: 10px;
  top: 10px;
  visibility: hidden;
  -webkit-animation: jiroXv 0.2s linear;
  animation: jiroXv 0.2s linear;
  -webkit-transition: visibility 0.2s linear;
  transition: visibility 0.2s linear;
  z-index: 1;
}

.c4 {
  font-family: Unica77LLWebRegular,Arial,serif;
  -webkit-font-smoothing: antialiased;
  font-size: 12px;
  line-height: 1.4em;
  margin-top: 10px;
  color: #666666;
  height: 16px;
}

<div
  className="c0"
>
  <div
    className="c1"
  >
    <label
      className="c2"
    >
      some label
    </label>
    <input
      className="c3"
      onBlur={[Function]}
      onChange={[Function]}
      onFocus={[Function]}
      value=""
    />
    <div>
      right side
    </div>
  </div>
  <div
    className="c4"
  >
    a note
  </div>
  <div
    className="c5"
  >
    an error
  </div>
</div>
`;
